feat: enable plugin allowlist

wep21 opened the pull request:
I enabled plugin blacklist for image transport publisher.
usage:
```
ros2 run usb_cam usb_cam_node_exe --ros-args -p image_raw.disable_pub_plugins:=["image_transport/compressed"]
```
```
❯ ros2 topic list
/camera_info
/image_raw
/image_raw/compressedDepth
/image_raw/theora
/parameter_events
/rosout
```

Commit:
feat: enable plugin blacklist

Signed-off-by: Daisuke Nishimatsu <border_goldenmarket@yahoo.co.jp>

wep21 commented:
@gbiggs @jacobperron @ijnek Could you review this PR? Also, I wolud like to backport this change into humble.

Commit:
use param base name instead of base topic

Signed-off-by: Daisuke Nishimatsu <border_goldenmarket@yahoo.co.jp>
Signed-off-by: wep21 <border_goldenmarket@yahoo.co.jp>

ijnek commented on image_transport/src/publisher.cpp:
```diff
@@ -110,9 +110,23 @@ Publisher::Publisher(
   impl_->base_topic_ = image_topic;
   impl_->loader_ = loader;
 
+  uint ns_len = node->get_effective_namespace().length();
```
Can ``param_base_name`` just be equal to ``base_topic``, rather than stripping off the namespace here?

wep21 replied:
@gbiggs @ijnek This comes from `image_transport_plugins`.
Which is better?

- use base topic name
  This looks like below.
 ```
/usb_cam:
  ros__parameters:
    image_raw:
      enable_pub_plugins:
      - image_transport/compressed
      format: jpeg
      jpeg_quality: 95
      png_level: 3
      tiff:
        res_unit: inch
        xdpi: -1
        ydpi: -1
```
- use whole topic name
  This looks like below.
```
/usb_cam:
  ros__parameters:
    image_raw:
      format: jpeg
      jpeg_quality: 95
      png_level: 3
      tiff:
        res_unit: inch
        xdpi: -1
        ydpi: -1
    /aaa/bbb/image_raw:
       enable_pub_plugins:
       - image_transport/compressed
```

ijnek replied:
Just the base topic name seems better.

ijnek commented:
As this seems like a revival of a feature that existed in ROS 1, I'd like the maintainers to decide on these changes. I do agree that a publisher getting created for every image transport plugin installed is annoying, but a whitelist sounds like a better idea than a blacklist.

wep21 commented:
friendly ping @gbiggs @jacobperron

gbiggs commented:
Please change this to use a whitelist rather than a blacklist, with an additional option to enable all publishers.

Commit:
create whitelist instead of blacklist

Signed-off-by: Daisuke Nishimatsu <border_goldenmarket@yahoo.co.jp>

Commit:
remove initial dot

Signed-off-by: Daisuke Nishimatsu <border_goldenmarket@yahoo.co.jp>

wep21 commented:
@gbiggs I changed this to whitelist at https://github.com/ros-perception/image_common/pull/264/commits/a6356ca4e6bd1bba3d2985ba505b7c38024374ff.

Commit:
apply format

Signed-off-by: Daisuke Nishimatsu <border_goldenmarket@yahoo.co.jp>

ijnek commented:
Should probably add all available transports to the whitelist is there are no plugins listed under ``enable_pub_plugins``. Otherwise, this change would break the default behavior of making all installed transports available (which many users depend on).

wep21 commented:
@ijnek okay, I will set all the plugins as a default parameter.

Commit:
set all transport names as default

Signed-off-by: Daisuke Nishimatsu <border_goldenmarket@yahoo.co.jp>

ijnek commented:
@wep21 Apologies for the delay. I'm just reviewing again, and I think there is no need for a separate std::set and a std::vector, if you simplify:

```cpp
  for (const auto & lookup_name : loader->getDeclaredClasses()) {
    const std::string transport_name = erase_last_copy(lookup_name, "_pub");
    if (whitelist.count(transport_name)) {
      try {
        auto pub = loader->createUniqueInstance(lookup_name);
        pub->advertise(node, image_topic, custom_qos, options);
        impl_->publishers_.push_back(std::move(pub));
      } catch (const std::runtime_error & e) {
        RCLCPP_ERROR(
          impl_->logger_, "Failed to load plugin %s, error string: %s\n",
          lookup_name.c_str(), e.what());
      }
    }
  }
```

to:

```cpp
for (transport_name : whitelist_vec) {
  try {
    auto pub = loader->createUniqueInstance(transport_name);
    pub->advertise(node, image_topic, custom_qos, options);
    impl_->publishers_.push_back(std::move(pub));
  } catch (const std::runtime_error & e) {
    RCLCPP_ERROR(
      impl_->logger_, "Failed to load plugin %s, error string: %s\n",
      transport_name.c_str(), e.what());
  }
}
```

Please check that these changes do work, as I haven't tested it.

Commit:
address review

Signed-off-by: Daisuke Nishimatsu <border_goldenmarket@yahoo.co.jp>

wep21 commented:
@ijnek I addressed your review [here](https://github.com/ros-perception/image_common/pull/264/commits/5a924947a24608a104aeaac8e6b6a07ef051b885). Could you review my changes? cc @gbiggs

gbiggs commented on image_transport/src/publisher.cpp:
```diff
@@ -110,19 +110,36 @@ Publisher::Publisher(
+      param_base_name +
+      ".enable_pub_plugins").get_value<std::vector<std::string>>();
+  }
+  for (size_t i = 0; i < whitelist_vec.size(); ++i) {
```
What is the purpose of copying the contents of the vector into a set if you're then just going to iterate over it?

wep21 replied:
@gbiggs This is from original code, but I guess this removes duplicated plugins declared in the parameters.

gbiggs commented on image_transport/src/publisher.cpp:
```diff
@@ -110,19 +110,36 @@ Publisher::Publisher(
+    param_base_name = param_base_name.substr(1);
   }
-
+  std::vector<std::string> whitelist_vec;
```
Please use the term "allowlist" instead of "whitelist".

wep21 replied:
@gbiggs I addressed [here](https://github.com/ros-perception/image_common/pull/264/commits/f24fa62231bb5623a6639560a8fd388de130e580)

wep21 committed:
rename whitelist to allowlist

Signed-off-by: wep21 <daisuke.nishimatsu1021@gmail.com>

ahcorde reviewed (approved):
@gbiggs can we merge this ?

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=19161)](http://ci.ros2.org/job/ci_linux/19161/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=13730)](http://ci.ros2.org/job/ci_linux-aarch64/13730/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=19910)](http://ci.ros2.org/job/ci_windows/19910/)

ahcorde reviewed (changes requested):
can you take a look to this windows failures ? https://ci.ros2.org/job/ci_windows/19910/console

ahcorde commented on image_transport/src/publisher.cpp:
```diff
@@ -110,19 +110,36 @@ Publisher::Publisher(
-  // nh.getParam(impl_->base_topic_ + "/disable_pub_plugins", blacklist_vec);
-  for (size_t i = 0; i < blacklist_vec.size(); ++i) {
-    blacklist.insert(blacklist_vec[i]);
+  uint ns_len = node->get_effective_namespace().length();
```
```suggestion
  auto ns_len = node->get_effective_namespace().length();
```

wep21 committed:
change type of namespace length

Co-authored-by: Alejandro Hernández Cordero <ahcorde@gmail.com>

wep21 commented:
@ahcorde Thank you for suggesting the fix. Could you run ci again?

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=19165)](http://ci.ros2.org/job/ci_linux/19165/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=13734)](http://ci.ros2.org/job/ci_linux-aarch64/13734/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=19916)](http://ci.ros2.org/job/ci_windows/19916/)

Sushant-Chavan commented:
> @gbiggs @jacobperron @ijnek Could you review this PR? Also, I wolud like to backport this change into humble.

@wep21 Thank you for the PR. Do you also plan to merge these changes into `humble`?

chrbrcknr commented:
Yes, it would be great to have this in `humble` too!

juanluishortelano commented:
Yes, would also love to have this on humble

noyzzz commented:
If you just want for example the compressed type you can enable it like this:

    start_gazebo_ros_image_bridge_back_cmd = Node(
        package='ros_gz_image',
        executable='image_bridge',
        name='ros_gz_image_bridge_back',
        arguments=['/camera_back/image_raw'],
        parameters=[{'camera_back.image_raw.enable_pub_plugins': ['image_transport/compressed']}],
        output='screen',
    )
note: for disabling everything except the raw one, you can't give an empty list, you just have to uninstall the image transport packages
